src/ray/core_worker/core_worker.cc

```diff
@@ -379,45 +379,54 @@ Status CoreWorker::SetClientOptions(std::string name, int64_t limit_bytes) {
   return plasma_store_provider_->SetClientOptions(name, limit_bytes);
 }
 
-Status CoreWorker::Put(const RayObject &object, ObjectID *object_id) {
+Status CoreWorker::Put(const RayObject &object,
+                       const std::vector<ObjectID> &contained_object_ids,
+                       ObjectID *object_id) {
   *object_id = ObjectID::ForPut(worker_context_.GetCurrentTaskID(),
                                 worker_context_.GetNextPutIndex(),
                                 static_cast<uint8_t>(TaskTransportType::RAYLET));
   reference_counter_->AddOwnedObject(*object_id, GetCallerId(), rpc_address_);
-  RAY_RETURN_NOT_OK(Put(object, *object_id));
-  // Tell the raylet to pin the object **after** it is created.
+  RAY_RETURN_NOT_OK(Put(object, contained_object_ids, *object_id));
   RAY_CHECK_OK(local_raylet_client_->PinObjectIDs(rpc_address_, {*object_id}));
   return Status::OK();
 }
 
-Status CoreWorker::Put(const RayObject &object, const ObjectID &object_id) {
+Status CoreWorker::Put(const RayObject &object,
+                       const std::vector<ObjectID> &contained_object_ids,
+                       const ObjectID &object_id) {
   RAY_CHECK(object_id.GetTransportType() ==
             static_cast<uint8_t>(TaskTransportType::RAYLET))
       << "Invalid transport type flag in object ID: " << object_id.GetTransportType();
+  // TODO(edoakes,swang): add contained object IDs to the reference counter.
+  // Tell the raylet to pin the object **after** it is created.
   return plasma_store_provider_->Put(object, object_id);
 }
 
 Status CoreWorker::Create(const std::shared_ptr<Buffer> &metadata, const size_t data_size,
+                          const std::vector<ObjectID> &contained_object_ids,
                           ObjectID *object_id, std::shared_ptr<Buffer> *data) {
   *object_id = ObjectID::ForPut(worker_context_.GetCurrentTaskID(),
                                 worker_context_.GetNextPutIndex(),
                                 static_cast<uint8_t>(TaskTransportType::RAYLET));
-  return Create(metadata, data_size, *object_id, data);
+  RAY_RETURN_NOT_OK(Create(metadata, data_size, contained_object_ids, *object_id, data));
+  if (data) {
```

Comment:
What is this check for?

Reply:
Ah this is a bit ugly but this is how we check if the object was already created (we should obviously return an error and not reach here instead). I will add a comment.